### PR TITLE
Add docs for perma-failed Deployments

### PR DESCRIPTION
**Pull request by kargakis**
cc: @janetkuo @kubernetes/docs 

Docs for https://github.com/kubernetes/features/issues/122

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1699)

### CODE REVIEW

File: docs/user-guide/deployments.md [hunk 1]
```diff
@@ -454,6 +454,125 @@ nginx-deployment-3066724191   0         0         1h
...
+Once the deadline is exceeded, meaning your Deployment will not report any progress for more than progressDeadlineSeconds,
+the deployment controller will add a Progressing condition with Status=False and Reason=ProgressDeadlineExceeded in the
+status of the Deployment.
+
+Note that at the moment, Kubernetes will do nothing more than just report with a `ProgressDeadlineExceeded` condition.
+In the future, once we support automatic rollbacks and users have opted-in, the deployment controller will rollback as
+soon as a failure condition is noticed.
+
+Another thing to note is transient errors eg. your rollout times out and a `ProgressDeadlineExceeded` condition is set to
+your Deployment because don't have enough quota. All you need is to scale down your Deployment, or scale down other Pods
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: You can't scale down pods. You can scale down other controllers or delete pods.

Comment (janetkuo) [inline docs/user-guide/deployments.md]: because don't have enough quota --> because of insufficient quota

File: docs/user-guide/deployments.md [hunk 2]
```diff
@@ -454,6 +454,125 @@ nginx-deployment-3066724191   0         0         1h
...
 
 
+## Failed Deployment
+
+There will be times your Deployment will get stuck trying to deploy its newest ReplicaSet without ever completing,
+due to insufficient quota, readiness probe failures, image pull error, insufficient permissions, limit ranges, or any
+application runtime misconfiguration. In those cases you can specify a timeout parameter in the spec
+(`spec.progressDeadlineSeconds`) that denotes the seconds you want to wait for your Deployment to report any progress.
+Once the deadline is exceeded, meaning your Deployment will not report any progress for more than progressDeadlineSeconds,
+the deployment controller will add a Progressing condition with Status=False and Reason=ProgressDeadlineExceeded in the
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Suggest quoting things like "Status=False" with `

File: docs/user-guide/deployments.md [hunk 3]
```diff
@@ -454,6 +454,125 @@ nginx-deployment-3066724191   0         0         1h
...
+soon as a failure condition is noticed.
+
+Another thing to note is transient errors eg. your rollout times out and a `ProgressDeadlineExceeded` condition is set to
+your Deployment because don't have enough quota. All you need is to scale down your Deployment, or scale down other Pods
+you may be running, or increase quota in your namespace, either by asking your administrator or by doing it yourself assuming
+you have the right permissions. Once the deployment controller notices that your Deployment is complete, it will transition
+the status of the condition into a succeeded one (Status=True, Reason=NewReplicaSetAvailable).
+
+A complete Deployment for the controller is one that: 
+* has minimum availability, meaning its available replicas are more than those required by the Deployment strategy (`status.
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: The bullet point isn't rendered well. I guess adding a new line will fix it

Comment (janetkuo) [inline docs/user-guide/deployments.md]: Remove the whitespace after `status.`

File: docs/user-guide/deployments.md [hunk 4]
```diff
@@ -454,6 +454,125 @@ nginx-deployment-3066724191   0         0         1h
...
+availableReplicas >= spec.replicas - maxUnavailable` where `maxUnavailable` is the maximum unavailable replicas that a
+Deployment allows at any point in time. It depends on the deployment strategy: Recreate deployments don't use it whereas
+Rolling deployments specify it in `spec.strategy.rollingUpdate.maxUnavailable`).
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Rolling **Update** deployments

File: docs/user-guide/deployments.md [hunk 5]
```diff
@@ -454,6 +454,125 @@ nginx-deployment-3066724191   0         0         1h
...
+* has all of its replicas updated to the latest revision (`spec.replicas == status.updatedReplicas`).
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Can we make it more clear on what a complete Deployment means: for rolling update deployments, both points should be true; for recreate ones, only the second one needs to be true (the first one doesn't apply)

File: docs/user-guide/deployments.md [hunk 6]
```diff
@@ -454,6 +454,125 @@ nginx-deployment-3066724191   0         0         1h
...
+
+Note that clients should always take into account generation/observedGeneration for ensuring that the latest state of a
+Deployment has been observed by the controller. For example, `kubectl rollout status` expects the Deployment generation
+to equal the observed generation (`d.generation <= d.status.observedGeneration`) before reporting any information provided
+in the status of the Deployment.
+
+Progress for a Deployment is:
+* the creation of the new replica set.
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: same here for the bullet point render issue

File: docs/user-guide/deployments.md [hunk 7]
```diff
@@ -454,6 +454,125 @@ nginx-deployment-3066724191   0         0         1h
...
+
+To make the controller report lack of progress for a Deployment after 10 minutes:
+
+```shell
+$ kubectl patch deployment/nginx-deployment -p '{"spec":{"progressDeadlineSeconds":600}}'
+"nginx-deployment" patched
+```
+
+If you describe the Deployment, you will notice a Conditions section:
+```
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: This part isn't rendered well either. Maybe a newline before ``` would fix it

File: docs/user-guide/deployments.md [hunk 8]
```diff
@@ -454,6 +454,125 @@ nginx-deployment-3066724191   0         0         1h
...
+  Type            Status  Reason
+  ----            ------  ------
+  Available       True    MinimumReplicasAvailable
+  Progressing     True    NewReplicaSetAvailable
+  ReplicaFailure  True    FailedCreate
+```
+
+This is how the status of the Deployment looks like if you would run `kubectl get deployment nginx-deployment -o yaml`
+(spec of the object is omitted for brevity):
+```
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: render issue again

File: docs/user-guide/deployments.md [hunk 9]
```diff
@@ -454,6 +454,125 @@ nginx-deployment-3066724191   0         0         1h
 Note: You cannot rollback a paused Deployment until you resume it.
 
 
+## Failed Deployment
+
+There will be times your Deployment will get stuck trying to deploy its newest ReplicaSet without ever completing,
+due to insufficient quota, readiness probe failures, image pull error, insufficient permissions, limit ranges, or any
+application runtime misconfiguration. In those cases you can specify a timeout parameter in the spec
+(`spec.progressDeadlineSeconds`) that denotes the seconds you want to wait for your Deployment to report any progress.
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: link to the section below for `spec.progressDeadlineSeconds`

File: docs/user-guide/deployments.md [hunk 10]
```diff
@@ -556,6 +675,15 @@ the rolling update starts, such that the total number of old and new Pods do not
 the new Replica Set can be scaled up further, ensuring that the total number of Pods running
 at any time during the update is at most 130% of desired Pods.
 
+### Progress Deadline Seconds
+
+`.spec.progressDeadlineSeconds` is an optional field that specifies the number of seconds you want
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Suggest linking back to the "failed deployment" section somewhere

**Comment by janetkuo** [comment]
We can add sub-sections in "failed deployment"

File: docs/user-guide/deployments.md (hunk 9 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

File: docs/user-guide/deployments.md (hunk 2 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

File: docs/user-guide/deployments.md (hunk 1 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

File: docs/user-guide/deployments.md (hunk 3 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

File: docs/user-guide/deployments.md (hunk 5 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: The first point definitely applies for recreate deployments - in order to have a complete deployment you have to have all pods available, right?

File: docs/user-guide/deployments.md (hunk 6 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

File: docs/user-guide/deployments.md (hunk 7 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

File: docs/user-guide/deployments.md (hunk 10 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

File: docs/user-guide/deployments.md [hunk 11]
```diff
@@ -454,6 +454,133 @@ nginx-deployment-3066724191   0         0         1h
...
+There will be times your Deployment will get stuck trying to deploy its newest ReplicaSet without ever completing,
+due to insufficient quota, readiness probe failures, image pull error, insufficient permissions, limit ranges, or any
+application runtime misconfiguration. In those cases you can specify a timeout parameter in the spec
+([spec.progressDeadlineSeconds](#progress-deadline-seconds)) that denotes the seconds you want to wait for your Deployment
+to report any progress. Once the deadline is exceeded, meaning your Deployment will not report any progress for more than
+progressDeadlineSeconds, the deployment controller will add a condition with `Type=Progressing`, `Status=False`, and
+`Reason=ProgressDeadlineExceeded` in the status of the Deployment. For more information on status conditions, see
+[here](https://github.com/kubernetes/kubernetes/blob/{{page.githubbranch}}/docs/devel/api-conventions.md#typical-status-properties).
+
+Note that at the moment, Kubernetes will do nothing more than just report with a condition with `Reason=ProgressDeadlineExceeded`.
```
Comment (jaredbhatti) [inline docs/user-guide/deployments.md]: We do not talk about "future features" in our technical docs.  It makes tracking changes really hard. Please change this paragraph, or at least remove the sentence starting with "In the future..."

File: docs/user-guide/deployments.md [hunk 12]
```diff
@@ -454,6 +454,133 @@ nginx-deployment-3066724191   0         0         1h
 Note: You cannot rollback a paused Deployment until you resume it.
 
 
+## Failed Deployment
+
+There will be times your Deployment will get stuck trying to deploy its newest ReplicaSet without ever completing,
```
Comment (jaredbhatti) [inline docs/user-guide/deployments.md]: Reword to: 

"Your Deployment may get stuck trying to deploy its newest ReplicaSet without ever completing, due to:
* Insufficient quota
* readiness probe failures
* image pull error
* insufficient permissions
* limit ranges
* application runtime misconfigurations

File: docs/user-guide/deployments.md [hunk 13]
```diff
@@ -454,6 +454,133 @@ nginx-deployment-3066724191   0         0         1h
...
+
+Note that at the moment, Kubernetes will do nothing more than just report with a condition with `Reason=ProgressDeadlineExceeded`.
+In the future, once we support automatic rollbacks and users have opted-in, the deployment controller will rollback as
+soon as a failure condition is noticed.
+
+Another thing to note is transient errors eg. your rollout times out and a condition with `Reason=ProgressDeadlineExceeded`
+is set to your Deployment because of insufficient quota. All you need is to scale down your Deployment, or scale down other
+controllers you may be running, or increase quota in your namespace, either by asking your administrator or by doing it yourself
+assuming you have the right permissions. Once the deployment controller notices that your Deployment is complete, it will
+transition the status of the Progressing condition into a succeeded one (`Status=True` and `Reason=NewReplicaSetAvailable`).
```
Comment (jaredbhatti) [inline docs/user-guide/deployments.md]: The sentence "All you need..." is very difficult to read.  Suggest breaking it into two: "To fix this issue, scale down your Deployment, scale down other controllers you may be running, or increase quota in your namespace.  You can do this by either by asking your administrator or doing it yourself assuming you have the right permissions."

File: docs/user-guide/deployments.md [hunk 14]
```diff
@@ -454,6 +454,133 @@ nginx-deployment-3066724191   0         0         1h
...
+
+A complete Deployment for the controller is one that: 
+
+* has minimum availability, meaning its available replicas are more than those required by the Deployment strategy 
+(`status.availableReplicas >= spec.replicas - maxUnavailable` where `maxUnavailable` is the maximum unavailable replicas
+that a Deployment allows at any point in time. It depends on the deployment strategy: Recreate deployments don't use it whereas
```
Comment (jaredbhatti) [inline docs/user-guide/deployments.md]: Remove "that" in "that a deployment allows at any point in  time"

**Comment by jaredbhatti** [comment]
Please fix the open docs issues.

File: docs/user-guide/deployments.md (hunk 11 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

File: docs/user-guide/deployments.md (hunk 12 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

File: docs/user-guide/deployments.md (hunk 13 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: ok

**Comment by kargakis** [comment]
@jaredbhatti thanks for the review, addressed all of your comments.

File: docs/user-guide/deployments.md [hunk 15]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+* application runtime misconfiguration
+
+In those cases you can specify a timeout parameter in the spec ([spec.progressDeadlineSeconds](#progress-deadline-seconds))
+that denotes the seconds you want to wait for your Deployment to report any progress. Once the deadline is exceeded, meaning
+your Deployment will not report any progress for more than progressDeadlineSeconds, the deployment controller will add a
+condition with `Type=Progressing`, `Status=False`, and `Reason=ProgressDeadlineExceeded` in the status of the Deployment.
+For more information on status conditions, see [here](https://github.com/kubernetes/kubernetes/blob/{{page.githubbranch}}/docs/devel/api-conventions.md#typical-status-properties). 
+Note that at the moment, Kubernetes will do nothing more than just report with a condition with `Reason=ProgressDeadlineExceeded`.
+
+Another thing to note is transient errors eg. your rollout times out and a condition with `Reason=ProgressDeadlineExceeded`
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: I think you're trying to say is something like this:

"If you're experiencing transient errors or failures with your Deployments, such as a timeout error during rollout that results in the condition `Reason=ProgressDeadlineExceeded` being set, you may have insufficient quota available. You can address an issue of insufficient quota by scaling down your Deployment, by scaling down other controllers you might be running, or by increasing quota in your namespace."

File: docs/user-guide/deployments.md [hunk 16]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+is set to your Deployment because of insufficient quota. To fix this issue, scale down your Deployment, or scale down other
+controllers you may be running, or increase quota in your namespace. You can increase quota by either asking your administrator
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: Remove the sentence starting with "You can increase quota by..."; that's outside the scope of this document.

File: docs/user-guide/deployments.md [hunk 17]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+or doing it yourself assuming you have the right permissions. Once the deployment controller notices that your Deployment is
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: Change last sentence:

"If you satisfy the quota conditions and the deployment controller then completes the Deployment rollout, you'll see the Deployment's status update to a successful value (`Status=True` and `Reason=NewReplicaSetAvailable`)."

File: docs/user-guide/deployments.md [hunk 18]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+complete, it will transition the status of the Progressing condition into a succeeded one (`Status=True` and
+`Reason=NewReplicaSetAvailable`).
+
+A complete Deployment for the controller is one that:
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: This entire section on what constitutes a complete deployment doesn't appear to be necessary.

I think you can remove everything from line 487 to line 507.

File: docs/user-guide/deployments.md [hunk 19]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+* image pull error
+
+* insufficient permissions
+
+* limit ranges
+
+* application runtime misconfiguration
+
+In those cases you can specify a timeout parameter in the spec ([spec.progressDeadlineSeconds](#progress-deadline-seconds))
+that denotes the seconds you want to wait for your Deployment to report any progress. Once the deadline is exceeded, meaning
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: Also, put in a paragraph break after "progress"; you need to insert the text from line 509 afterward.

File: docs/user-guide/deployments.md [hunk 20]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+
+Progress for a Deployment is:
+
+* the creation of the new replica set.
+
+* scaling up new pods.
+
+* scaling down old pods.
+
+To make the controller report lack of progress for a Deployment after 10 minutes:
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: These lines from 509 to 513 are the most important part of the entire section. They need to be moved up to the top of the section, immediately after the word "progress" on line 474.

File: docs/user-guide/deployments.md [hunk 21]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+
+```shell
+$ kubectl patch deployment/nginx-deployment -p '{"spec":{"progressDeadlineSeconds":600}}'
+"nginx-deployment" patched
+```
+
+If you describe the Deployment, you will notice a Conditions section:
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: This section is about how to detect and deal with a stalled Deployment. It isn't necessary to explain what a successful deployment looks like first. Remove everything from between lines 516 and 530.

File: docs/user-guide/deployments.md [hunk 22]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+  Progressing   True    NewReplicaSetAvailable
+```
+
+`Type=Available` with `Status=True` means that your Deployment has minimum availability. Minimum availability is dictated
+by the parameters specified in the deployment strategy. `Type=Progressing` with `Status=True` means that your Deployment
+is either in the middle of a rollout and it is progressing or that it has successfully completed its progress and the minimum
+required new replicas are available (see the Reason of the condition for the particulars - in our case
+`Reason=NewReplicaSetAvailable` means that the deployment is complete).
+
+For example, if you have insufficient pod quota and your Deployment tries to create more pods than allowed then
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: Move everything from line 532 to line 541 up, above line 477.

File: docs/user-guide/deployments.md [hunk 23]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+
+In those cases you can specify a timeout parameter in the spec ([spec.progressDeadlineSeconds](#progress-deadline-seconds))
+that denotes the seconds you want to wait for your Deployment to report any progress. Once the deadline is exceeded, meaning
+your Deployment will not report any progress for more than progressDeadlineSeconds, the deployment controller will add a
+condition with `Type=Progressing`, `Status=False`, and `Reason=ProgressDeadlineExceeded` in the status of the Deployment.
+For more information on status conditions, see [here](https://github.com/kubernetes/kubernetes/blob/{{page.githubbranch}}/docs/devel/api-conventions.md#typical-status-properties). 
+Note that at the moment, Kubernetes will do nothing more than just report with a condition with `Reason=ProgressDeadlineExceeded`.
+
+Another thing to note is transient errors eg. your rollout times out and a condition with `Reason=ProgressDeadlineExceeded`
+is set to your Deployment because of insufficient quota. To fix this issue, scale down your Deployment, or scale down other
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: Paragraph break after "quota." Insert the following:

"You can check to see if your Deployment has encountered quota problems by inspecting the Deployment with `kubectl get Deployment`."

Then move everything from line 544 to 572 underneath that sentence.

File: docs/user-guide/deployments.md [hunk 24]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+      object-counts, requested: pods=1, used: pods=3, limited: pods=2'
+    reason: FailedCreate
+    status: "True"
+    type: ReplicaFailure
+  observedGeneration: 3
+  replicas: 2
+  unavailableReplicas: 2
+```
+
+Eventually, once the Deployment progress deadline is exceeded, the status and reason of the Progressing condition
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: Everything from line 575 to line 591 needs to be moved up to line 475.

File: docs/user-guide/deployments.md (hunk 15 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: Insufficient quota is just an example of a transient error so that should be in the place of "a timeout error" in your suggestion.

File: docs/user-guide/deployments.md (hunk 18 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: I would prefer to keep this section around since we mention complete Deployments around w/o explaining anywhere what this really means.

**Comment by kargakis** [comment]
@devin-donnelly thanks for the review, I think I have addressed most if not all of your comments, ptal

File: docs/user-guide/deployments.md [hunk 25]
```diff
@@ -454,6 +454,142 @@ nginx-deployment-3066724191   0         0         1h
...
+  Progressing   True    NewReplicaSetAvailable
+```
+
+`Type=Available` with `Status=True` means that your Deployment has minimum availability. Minimum availability is dictated
+by the parameters specified in the deployment strategy. `Type=Progressing` with `Status=True` means that your Deployment
+is either in the middle of a rollout and it is progressing or that it has successfully completed its progress and the minimum
+required new replicas are available (see the Reason of the condition for the particulars - in our case
+`Reason=NewReplicaSetAvailable` means that the deployment is complete).
+
+A complete Deployment is one that:
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: I still don't think this section belongs here -- Everything between lines 567-585. Why are we describing a complete deployment, and deployment progress, in a section on failed deployments? I think this is either described elsewhere, or should be.

All you need is the last paragraph (line 587 to line 591) under a subheading called "Operating on a failed deployment".

**Comment by devin-donnelly** [comment]
One more structural change. It's much better now, thanks.

**Comment by devin-donnelly** [comment]
@janetkuo , are you doing the tech review on this one?

File: docs/user-guide/deployments.md (hunk 25 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: It seems that this section is the only one that mentions complete deployments. My reasoning for having it here was that you need to describe both concepts together in order to make them understood (what's a complete deployment, what's a failed deployment). That being said I am going to create a separate section for complete deployments prior to failed deployments.

File: docs/user-guide/deployments.md [hunk 26]
```diff
@@ -454,6 +454,154 @@ nginx-deployment-3066724191   0         0         1h
...
+
+A complete Deployment is one that: 
+
+* has minimum availability, meaning its available replicas are more than those required by the Deployment strategy 
+(`status.availableReplicas >= spec.replicas - maxUnavailable` where `maxUnavailable` is the maximum unavailable replicas
+that a Deployment allows at any point in time. It depends on the deployment strategy: Recreate deployments don't use it whereas
+Rolling Update deployments specify it in [spec.strategy.rollingUpdate.maxUnavailable](#max-unavailable)).
+* has all of its replicas updated to the latest revision (`spec.replicas == status.updatedReplicas`).
+
+Note that clients should always take into account generation/observedGeneration for ensuring that the latest state of a
```
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: I'm not sure what exactly this paragraph is trying to say. It needs some major revisions.

* Who are "clients" in this case? Users? If so, address the user directly as "you."
* What is generation/observedGeneration? Who provides the observedGeneration? Looks like a status field, so I guess Kubernetes provides that; what is generation?
* By "latest state of a Deployment has been observed by the controller", do you mean "What the user sees in kubectl actually reflects reality?" I'm having trouble parsing what this sentence means.

**Comment by devin-donnelly** [comment]
I like the new section organization; I have some revisions/comments to the text in the new sections.

File: docs/user-guide/deployments.md [hunk 27]
```diff
@@ -454,6 +454,154 @@ nginx-deployment-3066724191   0         0         1h
...
+* scaling up new pods.
+* scaling down old pods.
+
+You can monitor the progress for a Deployment by using `kubectl rollout status`.
+
+### Complete Deployment
+
+A complete Deployment is one that: 
+
+* has minimum availability, meaning its available replicas are more than those required by the Deployment strategy
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: "equals or exceeds" (>=, not >)

File: docs/user-guide/deployments.md (hunk 26 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: Client is anything that talks to the api server; a user that uses {kubectl, external client, directly via the API} or a controller (either internal, or external set up by an admin).

Comment (kargakis) [inline docs/user-guide/deployments.md]: > What is generation/observedGeneration? Who provides the observedGeneration? Looks like a status field, so I guess Kubernetes provides that; what is generation?

generation can be found in the metadata of an object, and some objects also implement observedGeneration in the status. Generation is updated every time the spec of the object is updated (ie. user changes something) and observedGeneration is updated by the respective controller of the object (controller observes the change that was done by the user).

> By "latest state of a Deployment has been observed by the controller", do you mean "What the user sees in kubectl actually reflects reality?" I'm having trouble parsing what this sentence means.

Once a client (again, either user or a controller) observes that metadata.generation == status.observedGeneration, then they can be sure that the current spec of the object has been observed by its controller).

Comment (kargakis) [inline docs/user-guide/deployments.md]: I removed this section in favor of a TODO

**Comment by devin-donnelly** [comment]
Thanks, @kargakis . I'm just going to do some minor cleanup and get it assigned for tech review, and we'll be ok to merge.

File: docs/user-guide/deployments.md [hunk 28]
```diff
@@ -454,6 +454,146 @@ nginx-deployment-3066724191   0         0         1h
...
+Your Deployment may get stuck trying to deploy its newest ReplicaSet without ever completing. This can occur due to some of the following factors:
+
+* Insufficient quota
+* Readiness probe failures
+* Image pull errors
+* Insufficient permissions
+* Limit ranges
+* Application runtime misconfiguration
+
+In those cases, you can specify a timeout parameter in the spec ([spec.progressDeadlineSeconds](#progress-deadline-seconds))
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: use `` to quote `spec.progressDeadlineSeconds`

File: docs/user-guide/deployments.md [hunk 29]
```diff
@@ -454,6 +454,146 @@ nginx-deployment-3066724191   0         0         1h
...
+
+### Complete Deployment
+
+Kubernetes marks a Deployment as _complete_ when it has the following characteristics: 
+
+* The Deployment has minimum availability. Minimum availability means that the Deployment's number of available replicas
+equals or exceeds the number required by the Deployment strategy.
+* All of the replicas associated with the Deployment have been updated to the latest version you've specified, meaning any
+updates you've requested have been completed.
+
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Add a line here for how to easily determine a Deployment is completed (such as from the output of `kubectl rollout status`)

File: docs/user-guide/deployments.md (hunk 28 above)
Comment (janetkuo) [inline docs/user-guide/deployments.md]: This sounds like when a Deployment failed, you need to specify a timeout for it. But we actually want the users to specify it beforehand. 

Rephrase it so that it's clear that users can use this timeout parameter to specify how the system should determine whether a Deployment is considered "failed to progress".

File: docs/user-guide/deployments.md [hunk 30]
```diff
@@ -454,6 +454,146 @@ nginx-deployment-3066724191   0         0         1h
...
+* Readiness probe failures
+* Image pull errors
+* Insufficient permissions
+* Limit ranges
+* Application runtime misconfiguration
+
+In those cases, you can specify a timeout parameter in the spec ([spec.progressDeadlineSeconds](#progress-deadline-seconds))
+that denotes the number of seconds to wait for your Deployment to report any progress.
+
+The following `kubectl` command sets the spec with `progressDeadlineSeconds` to make the controller report lack of progress for a Deployment after 10 minutes:
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: link to `progressDeadlineSeconds` below

File: docs/user-guide/deployments.md [hunk 31]
```diff
@@ -454,6 +454,146 @@ nginx-deployment-3066724191   0         0         1h
...
+
+```shell
+$ kubectl patch deployment/nginx-deployment -p '{"spec":{"progressDeadlineSeconds":600}}'
+"nginx-deployment" patched
+```
+
+Once the deadline has been exceeded, the Deployment controller adds a DeploymentCondition with the following attributes to
+the Deployment's status:
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Should we make it clear that it's under `.status.conditions`?

File: docs/user-guide/deployments.md [hunk 32]
```diff
@@ -454,6 +454,146 @@ nginx-deployment-3066724191   0         0         1h
...
+
+* Type=Progressing
+* Status=False
+* Reason=ProgressDeadlineExceeded
+
+See the [Kubernetes API conventions](https://github.com/kubernetes/kubernetes/blob/{{page.githubbranch}}/docs/devel/api-conventions.md#typical-status-properties) for more information on status conditions.
+
+Note that in version 1.5, Kubernetes will take no action on a stalled Deployment other than to report a status condition with
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: What does "take no action" mean? Please clarify it.

File: docs/user-guide/deployments.md [hunk 33]
```diff
@@ -454,6 +454,146 @@ nginx-deployment-3066724191   0         0         1h
...
+
+`Type=Available` with `Status=True` means that your Deployment has minimum availability. Minimum availability is dictated
+by the parameters specified in the deployment strategy. `Type=Progressing` with `Status=True` means that your Deployment
+is either in the middle of a rollout and it is progressing or that it has successfully completed its progress and the minimum
+required new replicas are available (see the Reason of the condition for the particulars - in our case
+`Reason=NewReplicaSetAvailable` means that the Deployment is complete).
+
+### Operating on a failed deployment
+
+All actions that apply to a complete Deployment also apply to a failed Deployment. You can scale it up/down, rollback
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: roll back

File: docs/user-guide/deployments.md [hunk 34]
```diff
@@ -454,6 +454,146 @@ nginx-deployment-3066724191   0         0         1h
...
+to a previous revision, or even pause it if you need to apply multiple tweaks in the Deployment pod template. Note
+that progress for a Deployment is not estimated while the Deployment is paused so you can safely pause a Deployment
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: "progress for a Deployment is not estimated" would you clarify this?

File: docs/user-guide/deployments.md [hunk 35]
```diff
@@ -556,6 +696,16 @@ the rolling update starts, such that the total number of old and new Pods do not
...
 at any time during the update is at most 130% of desired Pods.
 
+### Progress Deadline Seconds
+
+`.spec.progressDeadlineSeconds` is an optional field that specifies the number of seconds you want
+to wait for your Deployment to progress before the system reports back that the Deployment has
+[failed progressing](#failed-deployment) - surfaced as a condition with `Type=Progressing`, `Status=False`.
+and `Reason=ProgressDeadlineExceeded` in the status of the resource. The deployment controller will keep
+retrying the Deployment. In the future, once automatic rollback will be implemented, the deployment
+controller will rollback a Deployment as soon as it observes such a condition. If specified, this field
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: roll back

File: docs/user-guide/deployments.md [hunk 36]
```diff
@@ -556,6 +696,16 @@ the rolling update starts, such that the total number of old and new Pods do not
...
+needs to be greater than `.spec.minReadySeconds`.
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Is this line related to the previous line about auto rollback? If not, separate them (move auto rollback to the end); if so, make it clear that they're relevant

**Comment by devin-donnelly** [comment]
Hi @kargakis , would you mind having a look at Janet's feedback from a technical perspective? We're right up against our deadline for 1.5, and we need to get this merged ASAP. Thanks.

File: docs/user-guide/deployments.md (hunk 29 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: I have already added such a sentense in the previous section.

File: docs/user-guide/deployments.md (hunk 30 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: We link to there just above?

File: docs/user-guide/deployments.md (hunk 32 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: It will do nothing more than update the status.condition with type=Progressing to Status=False, Reason=ProgressDeadlineExceeded.

Comment (kargakis) [inline docs/user-guide/deployments.md]: Isn't "other than ..." clear enough?

File: docs/user-guide/deployments.md (hunk 34 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: The controller wont check if the Deployment failed.

File: docs/user-guide/deployments.md (hunk 36 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: No, it's not.

File: docs/user-guide/deployments.md (hunk 29 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: Added an additional line here that also points out the exit code of `rollout status` in case the Deployment completes.

**Comment by kargakis** [comment]
Hi @devin-donnelly @janetkuo I believe I have addressed most if not all comments, ptal

File: docs/user-guide/deployments.md (hunk 29 above)
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: Added grammar fix.

File: docs/user-guide/deployments.md (hunk 28 above)
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: Done in my own commit.

File: docs/user-guide/deployments.md (hunk 31 above)
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: Added code formatting.

File: docs/user-guide/deployments.md (hunk 34 above)
Comment (devin-donnelly) [inline docs/user-guide/deployments.md]: This actually belongs way higher in the document; it's kind of arbitrary here. Fixed in my own commit.

**Comment by kargakis** [comment]
@janetkuo when will these changes land in the master branch?

**Comment by janetkuo** [comment]
@kargakis yes we'll merge the change in 1.5 branch into master after 1.5 launch
